### PR TITLE
Change direction for perimeter extrusion at odd layers.

### PR DESCRIPTION
**Pull request by Noisyfox**
Initial test code based on supermerill/SuperSlicer@87245ae3c1df604fdff502c66a5cbdbf65464ab4

Test model from https://www.printables.com/model/235524-bridge-test

e74896754da6711d41f892315819e9e7
afb1a412df8b6cef4079f2b8ac748364
6f3dbde74cba8fe8df5f7e114a8537ff

### CODE REVIEW

**Comment by igiannakas** [comment]
Been testing this for a couple of prints together with remaining PRs from noisyfox. Looks good. Not sure what else is pending (and it’s in draft).

btw the new release is going to be so good! A super combo of features from prusa slicer & super slicer 🎉

**Comment by Noisyfox** [comment]
> Been testing this for a couple of prints together with remaining PRs from noisyfox. Looks good. Not sure what else is pending (and it’s in draft).
> 
> btw the new release is going to be so good! A super combo of features from prusa slicer & super slicer 🎉

I'm going to add filtering and config for this feature then the PR will be ready.

**Comment by igiannakas** [comment]
Stupid question but is there a particular reason why we wouldn’t want all loops to be alternating irrespective of the overhang angle? 

Also my 2 cents - I wouldn’t add this as a config option - it’s such an obvious improvement that it should always be active in my view.

**Comment by Noisyfox** [comment]
> Stupid question but is there a particular reason why we wouldn’t want all loops to be alternating irrespective of the overhang angle? Maybe seam inconsistency?
> 
> Also my 2 cents - I wouldn’t add this as a config option - it’s such an obvious improvement that it should always be active in my view.

To support modifier so that you can only turn this on at certain part of your model while not affecting the surface quality of other parts.

And alternating extrusion direction will affect surface quality, even for perfectly vertical surfaces, due to the fact that your nozzle /motor/belt/etc. behave slightly different on different directions:
e02c64b0a6c9c980aaf2faa5492f45dd
You can clear see the lower one has those alternating pattern on the side while the upper one does not.

**Comment by igiannakas** [comment]
> > Stupid question but is there a particular reason why we wouldn’t want all loops to be alternating irrespective of the overhang angle? Maybe seam inconsistency?
> > Also my 2 cents - I wouldn’t add this as a config option - it’s such an obvious improvement that it should always be active in my view.
> 
> To support modifier so that you can only turn this on at certain part of your model while not affecting the surface quality of other parts.
> 
> Alternating extrusion direction will affect surface quality, even for perfectly vertical surfaces, due to the fact that your nozzle /motor/belt/etc. behave slightly differently on different directions: e02c64b0a6c9c980aaf2faa5492f45dd You can clearly see the lower one has those alternating pattern on the side while the upper one does not.
> 
> So I want to allow user to fine control this feature to achieve the best of both worlds.

Thank you! In which case maybe an option to exclude the external perimeter from the reordering would be a good option to add. So to always avoid external perimeter artefacts but compromising a bit on the overhang quality.

 The extra perimeters on overhang should be able to deal with the majority of the overhang anyway.

**Comment by igiannakas** [comment]
Also why disable the algorithm for fuzzy skin?

**Comment by Sineos** [comment]
> In which case maybe an option to exclude the external perimeter from the change of direction would be a good option to add. So to always avoid external perimeter artefacts on all cases but compromising a bit on the overhang quality.

This would be a great feature to have.
As I'm printing a lot of ASA and also quite big objects, this alternating wall directions is a life saver (using Cura) since it significantly reduced warping.
Despite having around 50°C chamber temperature, I have seen 300+ mm ASA prints lifting the magnetic PEI sheet of its base.

**Comment by Noisyfox** [comment]
> Also why disable the algorithm for fuzzy skin?

Ah when I started working on this I reused the overhang detection result from the overhang slowdown code, which is disabled when fuzzy skin is used. Then later when I decided to not use it (as it's only available in classic overhang speed mode) and write my own steep overhang detect code, I just kept the same logic.

I think the logic behind this is when fuzzy skin is used, there will be countless of tiny overhangs due to the uneven surface, which could mess the algorithm up, and also since those overhangs are just so tiny they should not affect print quality.

I've never used this feature before personally, so I'm not sure if this will be a problem or not. Maybe if fuzzy skin is enabled we ignore the overhang detection and force reversing regardless? Since I think with the fuzzy skin you won't notice the surface quality decreasing cause by the alternating extrusion order?

**Comment by igiannakas** [comment]
> > Also why disable the algorithm for fuzzy skin?
> 
> Ah when I started working on this I reused the overhang detection result from the overhang slowdown code, which is disabled when fuzzy skin is used. Then later when I decided to not use it (as it's only available in classic overhang speed mode) and write my own steep overhang detect code, I just kept the same logic.
> 
> I think the logic behind this is when fuzzy skin is used, there will be countless of tiny overhangs due to the uneven surface, which could mess the algorithm up, and also since those overhangs are just so tiny they should not affect print quality.
> 
> I've never used this feature before personally, so I'm not sure if this will be a problem or not. Maybe if fuzzy skin is enabled we ignore the overhang detection and force reversing regardless? Since I think with the fuzzy skin you won't notice the surface quality decreasing cause by the alternating extrusion order?

Ah I see. Maybe let’s give it a try with it always enabled on fuzzy skin - it won’t show the artefacts due to the surface variation that’s inherent to it. Happy to test it for you (I’ve printed a few models with it - crochet style pumpkins for Halloween :D)

**Comment by igiannakas** [comment]
So i've been trying to test the new version but the app crashes on start up. Unfortunately no helpfull error logs other than an unknown exception has occured. Maybe something is wrong with the variable setup in the GUI. Any chance for a cross check?

**Comment by Noisyfox** [comment]
> So i've been trying to test the new version but the app crashes on start up. Unfortunately no helpfull error logs other than an unknown exception has occured. Maybe something is wrong with the variable setup in the GUI. Any chance for a cross check?

Sure, how could I reproduce the issue?

**Comment by Noisyfox** [comment]
> > > Also why disable the algorithm for fuzzy skin?
> > 
> > 
> > Ah when I started working on this I reused the overhang detection result from the overhang slowdown code, which is disabled when fuzzy skin is used. Then later when I decided to not use it (as it's only available in classic overhang speed mode) and write my own steep overhang detect code, I just kept the same logic.
> > I think the logic behind this is when fuzzy skin is used, there will be countless of tiny overhangs due to the uneven surface, which could mess the algorithm up, and also since those overhangs are just so tiny they should not affect print quality.
> > I've never used this feature before personally, so I'm not sure if this will be a problem or not. Maybe if fuzzy skin is enabled we ignore the overhang detection and force reversing regardless? Since I think with the fuzzy skin you won't notice the surface quality decreasing cause by the alternating extrusion order?
> 
> Ah I see. Maybe let’s give it a try with it always enabled on fuzzy skin - it won’t show the artefacts due to the surface variation that’s inherent to it. Happy to test it for you (I’ve printed a few models with it - crochet style pumpkins for Halloween :D)

I've added the support for fuzzy skin, feel free to give it a try.

**Comment by igiannakas** [comment]
> > So i've been trying to test the new version but the app crashes on start up. Unfortunately no helpfull error logs other than an unknown exception has occured. Maybe something is wrong with the variable setup in the GUI. Any chance for a cross check?
> 
> Sure, how could I reproduce the issue?
> 
> Are those crashes caused by those asserts?

Nope they wouldn’t - the assert statements only affect the debug build as that is when they are evaluated. Something else is going on - I’ll give a try to the latest build from softfevers pipeline as mine doesn’t seem to work with this PR and let you know.

**Comment by igiannakas** [comment]
@SoftFever tried your build of it- works great. Maybe some merge error on my side was causing it to crash.

File: src/libslic3r/PerimeterGenerator.cpp [hunk 1]
```diff
@@ -283,6 +332,22 @@ static ExtrusionEntityCollection traverse_loops(const PerimeterGenerator &perime
             BoundingBox bbox(polygon.points);
             bbox.offset(SCALED_EPSILON);
 
+            // Always reverse extrusion if use fuzzy skin: https://github.com/SoftFever/OrcaSlicer/pull/2413#issuecomment-1769735357
```
Comment (SoftFever) [inline src/libslic3r/PerimeterGenerator.cpp:335]: Question: why are we differenciate contour and hole for reversing?

Comment (Noisyfox) [reply src/libslic3r/PerimeterGenerator.cpp:335]: To improve surface quality when only contour/hole surface has steep overhang.

Comment (SoftFever) [reply src/libslic3r/PerimeterGenerator.cpp:335]: make sense 👍

**Review by SoftFever — APPROVED** [review]

**Comment by SoftFever** [comment]
@Noisyfox Thanks for the fantastic work.
Looks all good.
Merged 👍

**Comment by Rhotix** [comment]
So what is the use of this? I can print such overhangs flawless. Is there a issue to it or some special design? Please dont implent things as always active, since it can affect others setup

**Comment by Noisyfox** [comment]
> I can print such overhangs flawless.

I cannot.



> Please dont implement things as always active

It's not.

**Comment by vovodroid** [comment]
> this alternating wall directions is a life saver

@Sineos , what is infill percent in you prints? Interesting, that Cure not just alternate directions each layer, but also alternate each perimeter direction. Questions is whether it's necessary to reduce warping, or each layer alternating is good enough?

**Comment by vovodroid** [comment]
Hi, actually using term "odd layer" is wrong from user point of view. Though in code expression `layer_id % 2 == 1` is used, layer numeration in GUI starts from 1, i.e. layer_id+1. Thus reversing happens on even layers, starting from layer 2.

**Comment by Sineos** [comment]
> what is infill percent in you prints?

Typically 40% to 60%.
Interesting information that Cura is doing more here. Matches to my empirical feeling that Cura prints with this option are less prone to warping, but OTH it is difficult to tell since warping heavily depends on size and geometry of the part.

**Comment by vovodroid** [comment]
> Matches to my empirical feeling that Cura prints with this option are less prone to warping

Speed, layer height, seams position, etc. I successfully printed quite big ASA object sliced in Prusa with draft shield (not on photo):

image

**Pull request merged**